rust/oak/src/lib.rs

```diff
@@ -77,12 +83,64 @@ impl ReceiveChannelHalf {
     }
 }
 
-impl Read for ReceiveChannelHalf {
-    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
-        Ok(unsafe { wasm::channel_read(self.handle, buf.as_mut_ptr(), buf.len()) })
+impl ReceiveChannelHalf {
+    pub fn read_message(&mut self, buf: &mut Vec<u8>) -> std::io::Result<usize> {
+        let result = unsafe { wasm::channel_read(self.handle, buf.as_mut_ptr(), buf.capacity()) };
+        if result >= 0 {
+            unsafe {
+                buf.set_len(result as usize);
+            };
+            return Ok(result as usize);
+        }
+
+        // Extend the vector to be large enough for the message
+        debug!("Got {}, need {}", buf.len(), -result);
+        let msg_size = -result;
+        let extra = msg_size - (buf.len() as i32);
+        if extra < 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "Internal error: provided {} bytes for receive, asked for {}",
+                    buf.len(),
+                    msg_size
+                ),
+            ));
+        }
+        buf.reserve(extra as usize);
+
+        let result = unsafe { wasm::channel_read(self.handle, buf.as_mut_ptr(), buf.capacity()) };
```

Comment:
Consider doing this in a loop of some sort (even though it will execute either once or twice)?